accepted/00000-http-api.md

```diff
@@ -0,0 +1,220 @@
+
+
+# Detailed Design
+
+The design aims to provide a plain HTTP endpoint to the existing methods by reusing the available business logic with minimum redundancy. The bulk of the effort is to decouple the existing layers from protocol-dependent parts as much as possible, so that the most of the workflow can be utilized for both interfaces in an agnostic manner.
+
+![Current workflow](images/http-api-before.png)
+*Current workflow*
+
+The components marked with red are tightly coupled to the XMLRPC-specific functionality, and cannot be reused with HTTP as-is.
+
+The main goal is to restructure the workflow so that the agnostic parts can be isolated into separete components and be shared between two interfaces. These components are drawn in green in the proposed model below. This is especially important for handler classes, which are the linear growing components of the API implementation.
+
+To implement the plain HTTP over JSON interface, the existing [Spark framework](https://sparkjava.com) shall be used.
+
+![Proposed workflow](images/http-api-after.png)
+*Proposed workflow*
+
+The design is split into a number of areas that represent different aspects of the new API. The rest of the section describes each aspect in detail.
+
+1. [Endpoint registration](#1-endpoint_registration)
+2. [Authentication](#2-authentication)
+3. [I/O](#3-io)
+4. [Exception handling](#4-exception-handling)
+5. [Security](#5-security)
+6. [Documentation](#6-documentation)
+
+
+## 1. Endpoint Registration
+
+The most of the boilerplate code involves registration of the endpoint handlers, dispatching requests, and parameter matching. The XMLRPC framework provides an interface to register a class as a handler for a specific namespace. Further resolution of a method in a namespace is handled by the `BaseHandler` class using Java reflection to match the method names and parameters. `BaseHandler` is the base class that all the namespace handler classes inherit from.
+
+On the HTTP side, Spark allows registration of an endpoint as a "route" per each endpoint method, optionally including a number of named route parameters that construct the route path.
+
+The actual handler methods are designed to use POJOs as parameters. However, Spark expects a specific signature that takes `request` and `response` as parameters for route handlers. A simple wrapper shall unpack the parameters from the request and call the appropriate handler method for each request.
+
+A typical resource-oriented REST API relies on various HTTP verbs to provide contextual information on what type of operation an endpoint is used for. However, the XMLRPC protocol is designed around methods, rather than resources. Therefore, HTTP verbs will be restricted to GET for read-only, and POST for read-write operations regardless of whether they require a payload or not.
```

Comment:
I should've checked this out at the very beginning :see_no_evil: but here's what I've found:

We have 6 endpoints that have simple lists in input (mostly for system IDs).

We have only one that has a map:
```java
SnapshotHandler.listSnapshots(User user, Integer sid, Map dateDetails)
```
`dateDetails` contains only 2 constant optional keys: `startDate` and `endDate`

We're not doomed, however, since query strings (also Spark) actually support these data types. Here's what it would look like:

List example:
```
GET https://uyuni.server/rhn/manager/api/system/listLatestAvailablePackage?name=mypackage&systemIds[]=1001&systemIds[]=1002 (and so on)
```

Map example:
```
GET https://uyuni.server/rhn/manager/api/system/provisioning/snapshot/listSnapshots?sid=1001&dateDetails[startDate]=2022-02-27&dateDetails[endDate]=2022-02-28
```

Reply:
Solved for lists: https://github.com/uyuni-project/uyuni/pull/4910/commits/ac9d9947866b2c08239c320359c77b9741eac80d

Maps will be a bit more complex, though.
Considering the super-specific single edge case of `SnapshotHandler.listSnapshots`: I wonder if we're allowed to break its signature to adapt it.

Reply:
How about:
 - deprecating this method
 - keeping it exposed via XMLRPC, but not exposing it to HTTP
 - creating a new method with a different signature (or just an overload, if possible) and expose it on both protocols

From a downstream perspective this all will happen in a major version, so it should not be a problem.